### ISSUE TITLE
No Exception Opened for siteGroundTemperatureBuildingSurface .get()

### ISSUE THREAD

**Issue by chriswmackey** [NREL/OpenStudio#2291]
Hello beloved OpenStudio team,

I have noticed this bug a couple of times where I try to use the .get() method to get a certain object in my model and my whole program running the script crashes, seemingly because there is no exception opened for the operation.  The most recent issue that I have is with getting the monthly ground temperatures object.

Trying to run this python code, I get my whole program crashing:

import OpenStudio
model = OpenStudio .Model()
site = OpenStudio .Model.getSite(model)
grndTemps = site.siteGroundTemperatureBuildingSurface().get()

I dug through the error log a while ago and found that the error was because of something in OpenStudio and not the program running the script.  Let me know if you would like me to dig through the error log again to find the specific error message.

Alternatively, if there is some way of getting the ground temperature object without having to use the .get() method, I would be very interested in knowing it.

Thank you all for this wonderful software,
-Chris

cc: @mostaphaRoudsari

**Comment by macumber** [comment]
You should read the section **OpenStudio Measures and the boost::optional Type** in the [Measure Writing Guide](http://nrel.github.io/OpenStudio-user-documentation/reference/measure_writing_guide/#measure-writing).  Basically, Model.getSite returns an OptionalGroundTemperatureBuildingSurface and you need to check it using `.empty?` or `.is_initialized` before calling get.

That being said, I don't know why [this line](https://github.com/NREL/OpenStudio/blob/develop/openstudiocore/src/utilities/core/CommonImport.i#L116) is not working to cause an exception to be thrown instead of a seg fault crash when deferencing an unset optional.

**Comment by macumber** [comment]
Seems to be working correctly in Ruby:

```
irb(main):002:0> a = OptionalDouble.new
irb(main):009:0> begin
irb(main):010:1* a.get
irb(main):011:1> rescue Exception => e
irb(main):012:1> puts e.message
irb(main):013:1> end
Optional not initialized
```

What language are you using?  Are you sure you are trying to catch the exception?  Uncaught exceptions will crash the program.

**Comment by mbadams5** [comment]
@macumber He said in his original issue that he is using Python. Thus that exception will not be triggered because it is a ruby specific addition. That said, we could add similar functionality for the python bindings.

**Comment by mostaphaRoudsari** [comment]
@macumber @mbadams5, Chris is using IronPython and C# bindings.

**Comment by mbadams5** [comment]
@macumber I am able to reproduce this with the python bindings and it is a crash and not an exception (I wrapped the call in a try/catch). However, when I run it in IRB it does not crash and does produce the correct runtime error.

```
RuntimeError: Optional not initialized
	from (irb):4:in `get'
	from (irb):4
	from /usr/local/bin/irb:12:in `<main>'
```

**Comment by chriswmackey** [comment]
@macumber and @mbadams5 ,

Thanks for your attention on this.  I'm glad that you were able to reproduce the error and I apologize if my wording of the issue could have been better.  As Mostapha said, bindings are C# and language is IronPython.  Wrapping my IronPython code in a try/except loop did not stop my program from crashing.  Let me know if there is any other information that would be helpful and thanks again.

-Chris

**Comment by chriswmackey** [comment]
Also, checking `.empty?` or `.is_initialized` before calling `get()` makes a lot of sense.  The examples in the measure writing guide are clear but there seems to be no setter for siteGroundTemperatureBuildingSurface in the [site class](https://openstudio-sdk-documentation.s3.amazonaws.com/cpp/OpenStudio-1.12.0-doc/model/html/classopenstudio_1_1model_1_1_site.html) and so I was unsure about how to assign a new ground temperature object once I create it (in the event that `.empty?` or `.is_initialized` returns False).

**Comment by macumber** [comment]
SiteGroundTemperatureBuildingSurface is a unique object so there are 0 or 1 instances of it in the model.  To get the current one or create it if it doesn't exist use:

`model.getSiteGroundTemperatureBuildingSurface`

That will always return a non-optional object that you can use.  

Sounds like we might have a problem in our wrapping of the exception classes?

**Comment by macumber** [comment]
@chriswmackey you can look in [this file](https://github.com/NREL/OpenStudio/blob/develop/openstudiocore/src/model/Model_Common_Include.i) to see where these methods are added to the model for SWIG (they don't exist in C++).

**Comment by chriswmackey** [comment]
@macumber ,

Thank you for pointing me towards that method for getting a non-optional ground temperature object.  That definitely addresses my immediate issue and, as a result, I have been able to take out a workaround in the Honeybee code:
https://github.com/mostaphaRoudsari/Honeybee/commit/a806e825e7bc7786bf7806a0145dd65b15707d15

I have experienced my program crashing when using the .get() method in two other locations in the SDK (both cases were with optional objects if I remember correctly).  So there seems to be an issue with the wrapping of the exception classes as you say.  This example was the one that took the least amount of code to replicate but I can try to replicate the other 2 and post them here if that would be helpful.  I'll also take a deeper look at the file you posted once I get the chance.

Thank you again,
-Chris
